File: sentimental-credit/credit.py
# ask user for input
def userinput():
    number = input("Number: ")
    return number


number = userinput()


# check the conditions of this input
def general_condition(number):
    while not number.isdigit():
        number = input("Number: ")
    return number


number = general_condition(number)


# implement the luhns algorithm for the checksum
def luhns_condition(number):

    number_length = len(number)
    number_length = int(number_length)
    i = number_length-2
    j = number_length-1
    k = 0
    multiply = []
    first = []

    for i in range(i, -1, -2):
        digit = int(number[i])
        product = digit * 2

        if product > 9:
            product -= 9
        multiply.append(product)

    for j in range(j, -1, -2):
        digit = int(number[j])
        first.append(digit)

    checksum = sum(multiply) + sum(first)
    return checksum


checksum = luhns_condition(number)


# credit card conditions
def credit_condition(number, checksum):
    if checksum % 10 != 0:
        print("INVALID")
        return

    number = str(number)

    if len(number) == 15 and number[:2] == "34" or number[:2] == "37":
        print("AMEX")
    elif len(number) == 16 and number[:2] >= "51" and number[:2] <= "55":
        print("MASTERCARD")
    elif number[:1] == "4" and len(number) == 13 or len(number) == 16 and number[:1] == "4":
        print("VISA")
    else:
        print("INVALID")


credit_condition(number, checksum)

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>

string number;
int first_two_digits;
int first_digit;
int all_digits;
int digit;
int multiply;
int sum_second = 0;
int sum_first = 0;
int total;
int checksum;

int main(void)
{
    number = get_string("Number?:\n");

    first_two_digits = (number[0] - '0') * 10 + (number[1] - '0');
    first_digit = (number[0] - '0');
    all_digits = strlen(number);

    for (int i = all_digits - 2; i >= 0; i -= 2)
    {
        digit = number[i] - '0';
        multiply = digit * 2;

        if (multiply >= 10)
        {
            sum_second += (multiply % 10) + (multiply / 10);
        }
        else
        {
            sum_second += multiply;
        }
    }

    for (int i = all_digits - 1; i >= 0; i -= 2)
    {
        digit = number[i] - '0';
        sum_first += digit;
    }

    total = sum_second + sum_first;
    checksum = total % 10;

    if (checksum != 0)
    {
        printf("INVALID\n");
    }
    else if ((first_two_digits == 34 || first_two_digits == 37) && all_digits == 15)
    {
        printf("AMEX\n");
    }
    else if ((first_two_digits == 51 || first_two_digits == 52 || first_two_digits == 53 ||
              first_two_digits == 54 || first_two_digits == 55) &&
             all_digits == 16)
    {
        printf("MASTERCARD\n");
    }
    else if ((first_digit == 4) && (all_digits == 13 || all_digits == 16))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}
